src/dmd/dmodule.d

```diff
@@ -1356,3 +1383,232 @@ struct ModuleDeclaration
+private bool includeImportedModuleCheck(ModuleComponentRange components)
+{
+    assert(global.params.includeImports);
+
+    createMatchNodes();
+    size_t nodeIndex = 0;
+    while (nodeIndex < matchNodes.dim)
+    {
+        //printf("matcher ");printMatcher(nodeIndex);printf("\n");
+        auto info = matchNodes[nodeIndex++];
+        if (info.depth <= components.totalLength())
+        {
+            size_t nodeOffset = 0;
+            for (auto range = components;;range.popFront())
+            {
+                if (range.empty || nodeOffset >= info.depth)
+                {
+                    // MATCH
+                    //printf("matcher ");printMatcher(nodeIndex - 1);
+                    //printf(" MATCHES module '");components.print();printf("'\n");
+                    return !info.isExclude;
+                }
+                if (!range.front.equals(matchNodes[nodeIndex + nodeOffset].id))
+                {
+                    break;
+                }
+                nodeOffset++;
+            }
+        }
+        //printf("matcher ");printMatcher(nodeIndex-1);
+        //printf(" does not match module '");components.print();printf("'\n");
+        nodeIndex += info.depth;
+    }
+    assert(nodeIndex == matchNodes.dim, "code bug");
+    return includeByDefault;
+}
+
+// Matching module names is done with an array of matcher nodes.
+// The nodes are sorted by "component depth" from largest to smallest
+// so that the first match is always the longest (best) match.
+private struct MatcherNode
+{
+    union
+    {
+        struct
+        {
+            ushort depth;
+            bool isExclude;
+        }
+        Identifier id;
+    }
+    this(Identifier id) { this.id = id; }
+    this(bool isExclude, ushort depth)
+    {
+        this.depth = depth;
+        this.isExclude = isExclude;
+    }
+}
+
+/**
+$(D includeByDefault) determines whether to include/exclude modules when they don't
+match any pattern. This setting changes depending on if the user provided any "include patterns".
+When a user provides at least one "include pattern" it likely means they only want to include modules
+that they've explicitly declared, however, if they provide no patterns or only provide
+"exclude patterns", then it is likely they want to include everything except modules
+that match one of their exclusions.
+
+Note that a user can always override this default behavior using the special '.' pattern.
+*/
+private __gshared bool includeByDefault = true;
+private __gshared Array!MatcherNode matchNodes;
+
+version(none)
+private void printMatcher(size_t matcherIndex)
+{
+    auto info = matchNodes[matcherIndex++];
+    if (info.isExclude)
+        printf("-");
+
+    if (info.depth == 0)
+        printf(".");
+    else
+    {
+        foreach (i; 0..info.depth)
+        {
+            if (i > 0) printf(".");
+            printf("%s", matchNodes[matcherIndex + i].id.toChars);
+        }
+    }
+}
+
+private void createMatchNodes()
```

Comment:
Please add a DDoc header for all functions that include a summary, `Params` and `Returns`.